### PR TITLE
BaSD-MS-26: Update interviews logic

### PR DESCRIPTION
**Pull request by NicoCagnina**
# Interview properties:

| Property  | Type | Required | 
| ------------- | ------------- | ------------- |
| idPostulant | idPostulant | ✔️ |
| idClient | idClient | ✔️ |
| idApplication | idApplication   | ✔️ |
| status | String  | ✔️  |
| date | Date  | ✔️ |
| notes  | String  | ❌ |

### Methods:

- listInterviews
- createInterview
- updateInterview
- deleteInterview

### Status:

- [x] Validations
- [x] Postman colection updated
- [x] Successful interaction with dB
- [ ] Successful interaction with other collections (not enough collections to test)

### CODE REVIEW

File: controllers/interviews.js [hunk 1]
```diff
@@ -1,71 +1,68 @@
...
+  const queries = req.query;
+  const interview = new Interviews({
+    idPostulant: queries.idPostulant,
+    idClient: queries.idClient,
+    idApplication: queries.idApplication,
+    status: queries.status,
+    date: queries.date,
+    notes: queries.notes,
+  });
+
```
Comment (ignacioaristo) [inline controllers/interviews.js:23]: Nico, i think we must use body instead of queries.

**Review by ignacioaristo — CHANGES_REQUESTED** [review]

File: validations/interviews.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  }
+  if (!bodyReq.application) {
+    return res.status(400).send('Application id is required');
+  }
+  if (!bodyReq.status) {
+    return res.status(400).send('Status is required');
+  }
+  if (!bodyReq.date) {
+    return res.status(400).send('Date is required');
+  }
```
Comment (estefrare) [inline validations/interviews.js:23]: also should validate the type of each value

File: controllers/interviews.js [hunk 3]
```diff
@@ -1,71 +1,68 @@
...
-        id: req.query.id,
-        idClient: req.query.idClient,
-        idPostulant: req.query.idPostulant,
-        idApplication: req.query.idApplication,
-        date: req.query.date
+  const bodyReq = req.body;
+  const interview = new Interviews({
+    postulant: req.body.postulant,
+    client: req.body.client,
+    application: bodyReq.application,
```
Comment (estefrare) [inline controllers/interviews.js:18]: use `bodyReq` or `req.body` for all cases

**Review by estefrare — CHANGES_REQUESTED** [review]

File: controllers/interviews.js [hunk 4]
```diff
@@ -1,68 +1,69 @@
...
+      postulant: bodyReq.postulant,
+      client: bodyReq.client,
+      application: bodyReq.application,
+      status: bodyReq.status,
+      date: bodyReq.date,
+      notes: bodyReq.notes,
+    },
+    { new: true },
+    (error, newInterview) => {
+      if (newInterview === null) {
```
Comment (estefrare) [inline controllers/interviews.js]: just write `if (!newInterview) {`

File: controllers/interviews.js [hunk 5]
```diff
@@ -1,68 +1,69 @@
...
+        return res.status(400).json('Interview id does not exist');
```
Comment (estefrare) [inline controllers/interviews.js]: status code should be 404

File: controllers/interviews.js [hunk 6]
```diff
@@ -1,68 +1,69 @@
...
 
 const deleteInterview = (req, res) => {
-  let deletedInterview;
-  const filteredInterview = Interviews.filter((interview) => {
-    if (interview.id === req.query.id) {
-      deletedInterview = interview;
-      return false;
+  Interviews.findByIdAndDelete(req.params.id, (error, pointedInterview) => {
+    if (pointedInterview === null) {
+      return res.status(400).json('Interview id does not exist');
```
Comment (estefrare) [inline controllers/interviews.js]: the same as above

File: controllers/interviews.js [hunk 7]
```diff
@@ -1,68 +1,69 @@
...
-  if (!deletedInterview) {
-    return res.status(404).send('Interview NOT found');
-  }
-  return fs.writeFile('./data/interviews.json', JSON.stringify(filteredInterview), {}, (error) => {
     if (error) {
-      return res.status(400).send(error);
+      return res.status(400).json(error);
     }
-    return res.status(204).send(deletedInterview);
+    return res.status(200).json('interview deleted successfully');
```
Comment (estefrare) [inline controllers/interviews.js]: 204

File: models/Interviews.js [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    required: true,
+    ref: 'Applications',
+  },
+  status:
+  {
+    type: String,
+    enum: ['succesful', 'failed', 'cancelled', 'assigned', 'confirmed'],
+    default: 'assigned',
+    required: true,
+  },
```
Comment (estefrare) [inline models/Interviews.js:33]: fix syntax
```
  status: {
    type: String,
    enum: ['succesful', 'failed', 'cancelled', 'assigned', 'confirmed'],
    default: 'assigned',
    required: true,
  },
  ```

**Comment by estefrare** [comment]
@NicoCagnina remember to delete the `interviews.json` file

File: validations/interviews.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    return res.status(400).send('Application id is wrong or missing');
+  }
+  if (!bodyReq.status || !bodyReq.status.match(enu)) {
+    return res.status(400).send('Status is wrong or missing');
+  }
+  if (!bodyReq.date || !bodyReq.date.match(/^\d{4}-(0[1-9]|1[0-2])-(0[1-9]|[12][0-9]|3[01])$/)) {
+    return res.status(400).send('Date is wrong or missing');
+  }
+  next();
+  return res.status(200);
```
Comment (estefrare) [inline validations/interviews.js]: why are you sending a response here?

**Review by estefrare — CHANGES_REQUESTED** [review]

**Comment by estefrare** [comment]
ESlint error
image

File: models/Interviews.js [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+  status: {
+    type: String,
+    enum: ['succesful', 'failed', 'cancelled', 'assigned', 'confirmed'],
+    default: 'assigned',
+    required: true,
+  },
+  date: { type: Date, required: true },
+  notes: { type: String },
+});
```
Comment (estefrare) [inline models/Interviews.js]: add `timestamps: true `

**Review by estefrare — CHANGES_REQUESTED** [review]

**Review by estefrare — APPROVED** [review]

**Review by ignacioaristo — APPROVED** [review]

**Pull request merged**